### ISSUE TITLE
SIGTRAP on startup attempt on 32-bit Debian Testing (trixie/sid)

### ISSUE THREAD

**Issue by avillacis** [LBCrion/sfwbar#154]
I have this 32-bit laptop (no 64-bit support) running Debian Testing, by installing bookworm then switching to testing in the repositories.
When running sway, and trying to run swfbar from the foot terminal, I get this:
```
21:09:48,78 module: register expr function 'mid'
21:09:48,78 module: register expr function 'replace'
21:09:48,78 module: register expr function 'pad'
21:09:48,78 module: register expr function 'extract'
21:09:48,78 module: register expr function 'time'
21:09:48,78 module: register expr function 'disk'
21:09:48,78 module: register expr function 'ActiveWin'
21:09:48,78 module: register expr function 'min'
21:09:48,78 module: register expr function 'max'
21:09:48,78 module: register expr function 'str'
21:09:48,78 module: register expr function 'val'
21:09:48,78 module: register expr function 'upper'
21:09:48,78 module: register expr function 'lower'
21:09:48,78 module: register expr function 'gtkevent'
21:09:48,78 module: register expr function 'widgetid'
21:09:48,78 module: register expr function 'windowinfo'
21:09:48,78 module: register expr function 'escape'
21:09:48,78 module: register expr function 'read'
21:09:48,78 module: register action 'Exec'
21:09:48,78 module: register action 'Function'
21:09:48,78 module: register action 'PipeRead'
21:09:48,78 module: register action 'MenuClear'
21:09:48,78 module: register action 'MenuItemClear'
21:09:48,78 module: register action 'Menu'
21:09:48,78 module: register action 'SwayCmd'
21:09:48,78 module: register action 'SwayWinCmd'
21:09:48,78 module: register action 'MpdCmd'
21:09:48,78 module: register action 'Config'
21:09:48,78 module: register action 'SetMonitor'
21:09:48,78 module: register action 'SetLayer'
21:09:48,78 module: register action 'SetMirror'
21:09:48,78 module: register action 'BlockMirror'
21:09:48,78 module: register action 'SetBarSize'
21:09:48,78 module: register action 'SetBarID'
21:09:48,78 module: register action 'SetExclusiveZone'
21:09:48,78 module: register action 'SetBarSensor'
21:09:48,78 module: register action 'SetBarVisibility'
21:09:48,78 module: register action 'SetValue'
21:09:48,78 module: register action 'SetStyle'
21:09:48,78 module: register action 'SetTooltip'
21:09:48,78 module: register action 'UserState'
21:09:48,78 module: register action 'PopUp'
21:09:48,78 module: register action 'ClientSend'
21:09:48,78 module: register action 'Focus'
21:09:48,78 module: register action 'Close'
21:09:48,78 module: register action 'Minimize'
21:09:48,78 module: register action 'Maximize'
21:09:48,78 module: register action 'UnMinimize'
21:09:48,78 module: register action 'UnMaximize'
21:09:48,78 module: register action 'Eval'
21:09:48,78 module: register action 'SwitcherEvent'
21:09:48,78 module: register action 'ClearWidget'
21:09:48,78 module: register action 'TaskbarItemDefault'
21:09:48,78 module: register action 'WorkspaceActivate'
21:09:48,81 default output: LVDS-1
21:09:48,82 app_id: 'foot', title 'foot'
21:09:48,82 app_id: 'foot', title 'foot'
21:09:48,82 app_id: 'pcmanfm', title 'alex'
21:09:48,82 app_id: 'chromium', title 'Bookmarks - Chromium'
21:09:49,28 include: sfwbar.config -> /home/alex/.config/sfwbar/sfwbar.config

** (sfwbar:13652): ERROR **: 21:09:49.286: Configuration file doesn't specify any features
```
Immediately after, the program terminates with a SIGTRAP.

My swfbar.confg file contains this:
```
Set Term = "foot"

function("SfwbarInit") {
  SetBarId "bar-0"
  SetLayer "top"
}

include("winops.widget")

layout "sfwbar" {
  include("startmenu.widget")

  button {
    style = "launcher"
    value = $Term
    action = Exec $Term
  }

  button {
    style = "launcher"
    value = "firefox"
    action = "firefox"
  }

  taskbar {
    rows = 1
    group = true
    group cols = 1
    group style = "taskbar_group"
    group labels = true
    group icons = true
    icons = true
    labels = true
    action[3] = Menu "winops"
  }

  label { css = "* { -GtkWidget-hexpand: true; min-height: 30x; }" }

  include("battery-svg.widget")

  label {
    value = Time("%k:%M")
    style ="clock"
  }
}

#CSS
@define-color lab_bg_color RGBA(0, 0, 0, 0.85);
@define-color lab_active_color #bc4b4f;
@define-color lab_hover_color RGBA(255, 255, 255, 0.1);
@define-color lab_text_color #d1d1d1;
@define-color lab_menu_bg_color #353535;

window {
  -GtkWidget-direction: bottom;
  background-color: @lab_bg_color;
}

label {
  font-size: 14px;
  color: @lab_text_color;
  text-shadow: none;
}

button,
button image {
  outline-style: none;
  box-shadow: none;
  background-image: none;
  border-image: none;
  border-radius: 0;
  min-height: 30px;
  min-width: 30px;
  margin: 0px;
  margin-right: 3px;
  border: 0px;
  background-color: RGBA(0, 0, 0, 0.0);
  -GtkWidget-valign: center;
}

button:hover {
  background-color: @lab_hover_color;
}

button#taskbar_group_normal,
button#taskbar_group_normal:hover,
button#taskbar_group_active {
  border: 0px;
  -GtkWidget-valign: center;
}

button#taskbar_group_active {
  background-color: @lab_active_color;
}

button#taskbar_group_active:hover {
  background-color: @lab_hover_color;
}

#menu_item,
#menu_item *,
#menu_item image,
#menu_item label {
  -GtkWidget-halign: start;
  color: white;
}

#menu_item image {
  min-width: 16px;
  min-height: 16px;
  padding-right: 2px;
  margin-right: 7px;
  margin-left: 3px;
}

menu {
  background-color: @lab_menu_bg_color;
}

menu arrow {
  background: none;
}

label#clock {
  color: @lab_text_color;
  -GtkWidget-vexpand: true;
  -GtkWidget-valign: center;
  font-size: 18px;
  margin-right: 8px;
}

```

I uninstalled the repo package and compiled from source up to commit a6748b3fdeaf76a926e2ea4bc096cecfda55fcb2, but the problem still persists.

I am in the process of installing a 64-bit virtual Debian bookworm machine on a separate system to test whether this is an unintended 32-bit incompatibility, or a different bug.

**Comment by LBCrion** [comment]
Interesting. I've never tested the code on the 32bit machine and I can
think of one place where I assume 64bit pointers (but this is specific to
hyprland support). It looks like the program fails to read the config file.
I assume the sfwbar.config is readable by the user running sfwbar?

**Comment by avillacis** [comment]
> Interesting. I've never tested the code on the 32bit machine and I can think of one place where I assume 64bit pointers (but this is specific to hyprland support). It looks like the program fails to read the config file. I assume the sfwbar.config is readable by the user running sfwbar?

```
ls -l .config/sfwbar/
total 4
-rw-r--r-- 1 alex alex 2187 feb 27 22:23 sfwbar.config

```
Permissions are correctly configured as shown above.

Just in case, I ran strace on the self-compiled sfwbar, and it shows that the configuration file is correctly located, opened and read into memory.

**Comment by avillacis** [comment]
On a 64-bit virtual machine, sfwbar runs correctly, so this really seems some sort of 32-bit incompatibility bug.

**Comment by LBCrion** [comment]
Ok, this will probably take some time to fix. I don't have any 32bit
hardware left, so will need to setup a 32bit vm and find a linux distro
supporting 32bit and then find all implicit 64bit assumptions in the code.
I know at least one place where I assume it explicitly, but as mentioned
before it shouldn't affect you,.

**Comment by avillacis** [comment]
I dedicated a few hours to diving into the config file parsing code, and I believe I have found the root cause. The issue is an invalid parameter value on the call to g_scanner_input_text() that results in a pointer wraparound in 32-bit targets. The fix is a one-liner, and the commit message is much longer than the fix. Pull request is https://github.com/LBCrion/sfwbar/pull/155 . After applying this fix, sfwbar starts correctly and displays its bar under sway.
